File: crates/r4s_engine/src/resolve.rs
use crate::dice::DiceRoll;
use crate::types::TieWinner;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Outcome {
    Success,
    Failure,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct RollOutcome {
    pub outcome: Outcome,
    pub player_total: u32,
    pub opposition_total: u32,
    pub tie_winner: TieWinner,
    /// 1 si la tirada falla, 0 si tiene éxito.
    pub xp_gained: u32,
}

/// Compara la tirada del jugador con la oposición del DM.
pub fn resolve(player: &DiceRoll, opposition: &DiceRoll, tie_winner: TieWinner) -> RollOutcome {
    let player_total = player.total();
    let opposition_total = opposition.total();
    let success =
        player_total > opposition_total || (player_total == opposition_total && tie_winner == TieWinner::Player);
    let outcome = if success { Outcome::Success } else { Outcome::Failure };
    RollOutcome {
        outcome,
        player_total,
        opposition_total,
        tie_winner,
        xp_gained: u32::from(outcome == Outcome::Failure),
    }
}

File: web/src/engine/resolve.ts
import type { DiceRoll } from './dice';
import type { TieWinner } from './types';

export type Outcome = 'success' | 'failure';

export interface RollOutcome {
  outcome: Outcome;
  playerTotal: number;
  oppositionTotal: number;
  tieWinner: TieWinner;
  /** 1 si la tirada falla, 0 si tiene éxito. */
  xpGained: number;
}

/** Compara la tirada del jugador con la oposición del DM: sus dados o un objetivo fijo (ya como total). */
export function resolve(player: DiceRoll, opposition: DiceRoll | number, tieWinner: TieWinner): RollOutcome {
  const playerTotal = player.total();
  const oppositionTotal = typeof opposition === 'number' ? opposition : opposition.total();
  const success = playerTotal > oppositionTotal || (playerTotal === oppositionTotal && tieWinner === 'player');
  const outcome: Outcome = success ? 'success' : 'failure';
  return { outcome, playerTotal, oppositionTotal, tieWinner, xpGained: outcome === 'failure' ? 1 : 0 };
}
